File: 383-ransom-note/383-ransom-note.js
/**
 * @param {string} ransomNote
 * @param {string} magazine
 * @return {boolean}
 */
const canConstruct = function(ransomNote, magazine) {
    const map = {};
    for (const c of magazine) {
        map[c] = ++map[c] || 1;
    }
    for (const c of ransomNote) {
        if (!map[c]--) return false;
    }
    return true;
};

File: 383-ransom-note/383-ransom-note.rb
# @param {String} ransom_note
# @param {String} magazine
# @return {Boolean}
def can_construct(ransom_note, magazine)
    have = magazine.split(//).tally
    ransom_note.each_char do |c|
        return false if [nil, 0].include? have[c]
        have[c] -= 1
    end
    true
end